README
This is the Beehive Monitoring System (BMS)
designed and created by Evangelos Rokas (c) 2015,2016
Continued major development during 2017
Continued major development during 2018


The BMS is a Arduino-compatible project that monitors the health
of beehives and sends data to an internet server.

Various modules have been used for the creation of the project.

EAGLE Schematic Capture System was used to design the schematics and create
the PCB files.

SeeedStudio website was used to fabricate the initial Rev1 PCBs which were
used to make the prototypes for the BMS.

Arduino Enviroment was used in initial stages in order to settle everything.

Linux operating system was used in the actual project development.

Arduino-Makefile is a project utilizing Arduino Enviroment in order to do
production development under Linux.  Arduino-Makefile project is a work of
others: This makefile was originally created by [Martin Oldfield]
(http://mjo.tc/atelier/2009/02/arduino-cli.html) and he maintained it till
v0.10.2.  From May 2013, it is maintained by [Sudar Muthu]
(http://hardwarefun.com/tutorials/compiling-arduino-sketches-using-makefile)
and [Simon John] (https://github.com/sej7278) with the help of [40+
contributors](https://github.com/sudar/Arduino-Makefile/graphs/contributors).


I2CDev library has been used in order to communicate with the various
modules of the project.



9/2/2017 Evangelos Rokas <evrokas@gmail.com>

This is the come back of the Beehive Monitoring System Project.  Basically,
after half a year of hibernation, I decided to come back to the project in
order to help my father with beekeeping.

RTC														is working
TEMPERATURE/HUMIDIty sensor		is working
ACCELERATOR sensor						is working (movement alert should be implemented)
GPS receiver									is working (software integration is needed to be written)
GSM/GPRS module								is partially working, working on the power of the module which
																is causing restarts even on normal voltage range(!)

After working prototype is ready, a new schematic will be created and new
PCBs will be ordered.


31/3/2018 Evangelos Rokas <evrokas@gmail.com>

As of March 2018 the GSM/GPRS module is actually working, and sending data
to an experimental website, running on Apache, Mysql and PHP.  Data were
gathered and presented in a table and graphical manner using various tools. 
Nevertheless, the website front-end is far from complete, and also the
GSM/GPRS firmware needs to be more stable.  Also, EEPROM data storing is not
complete.  EEPROM data storing is the work which is actually must be done
during this semester.



----- POST data -----

up vote
5
down vote
httpdata=100,10000 means that SIM800 should expect 100 bytes within 10 seconds.

This is how I accomplished this using the HTTP client:

AT+HTTPINIT
AT+HTTPPARA="CID",1
AT+HTTPPARA="URL","http://url.com/endPoint"
AT+HTTPPARA="CONTENT","application/json"
AT+HTTPDATA=40,10000
At this point, the SIM800 should respond with "DOWNLOAD". Which means it's expecting your data. Send in your data; in my case:

{"location_id": 238, "fill_percent": 90}
Wait 10 seconds to send the rest of the commands. Then:

AT+HTTPACTION=1
AT+HTTPREAD
AT+HTTPTERM
That did it for me. Hope it helps.

This is where I got the information from: http://www.raviyp.com/embedded/194-sim900-gprs-http-at-commands


	
answered Dec 9 '15 at 19:12

Alan Alvarez
936
add a comment
up vote
4
down vote
I managed to get it to do what I need, this code-snippet will likely help others

AT+CGATT=1
AT+CIPMUX=0
AT+CSTT="APN","USER","PASS"
AT+CIICR
AT+CIFSR # ip up - gprs working
AT+CIPSHUT # shut down 
# Now do a post request to remote server api in json format. Express server
AT+CIPSTART="TCP","IP_ADDR|DOMAIN","2000"
AT+CIPSEND=119 # Num of char in tcp/ip data, \r & \n count as 1 char
POST /post HTTP/1.1
Host: **.**.***.***
Content-Type: application/json
Content-Length:23

{"postkey":"postvalue"}
Hope this helps the next person stuck on it.


GET /data.php?action=getconf&nodeId=10200 HTTP/1.1
Host: erns.sytes.net:8088
User-Agent: beewatch-firmware/0.1 arduino/1.8.1


30-4-2018
TODO:
* Enable interrupt on change of Serial RX, so to wake up from sleep
when external module is attached to the serial port.

* when waiting for module to register, perform a 3 step procedure, initiate
a variable for 10 secs (this is the most likely the module will registerd),
if in this interval module is not registered, set a flag, and increase the
interval by 10 secs, and go to sleep, for a specified time frame, after
this, try again for network communication



09-06-2018
TODO:
Having a variable for repeating log (or net cycles), seems not to be enough. 
Whenever the module resets, this variable starts over again.  This creates
the problem, that the module might fire in regular but not specific times
(or intervals).

There are 2 solutions:
A.  Add a delay variable so when one reprograms the log/net cycles, the
module must wait for the new 'delay' variable so when it starts cycling in a
known pattern.

B.  Implement an algorithm similar to linux program 'crontab', which
specifies certain intervals (or specific minute/hour criteria) when to fire
log/net cycle.

In this implementation in order to reduce memory usage one can apply a 60
bits variable for minutes and a 24 bit variable for hours (one could add one
more 7 bit variable for days).  When a specific bit is set to '1' then the
event is allowed to fire, in the specific minute:hour(:day) time.  RAM usage
is: 60+24+7 = 91 bits (or 84 if days are not used) which means an extra 12
bytes of memory for each event (12 for Log, 12 for Net).



25-09-2018
Progress
The software is progressing from alpha to beta stage. Soon it will be ready
to do a field test on real enviroment.


* Sending data to the server via POST is ok, this reduced the time needed for
GPRS/GSM connection A LOT (!), it now needs almost 20-30 seconds to send a chunk
of data when it needed almost 200 seconds to send 17 records (!!!). This reduces
a lot battery consumption and increases life capacity.

* An error logging mechanism is began to be created, so errors during operations
are stored in RAM and sent to the server at each connection (or earlier perhaps).

* Maintenance mode is working good (but not perfect yet), you can actually execute
various commands.

* What is very important to implement is a fail safe mechanism that will restart the
module when something goes wrong. In such a case logged records can be lost, but it
is absolutely important for the module to resume normal operation.


** Note from the SIM900_AN_TCPIP_V100:
Error handling:
If en error occurs during TCP/UDP connection, it is suggested to close the connection
with AT+CIPCLOSE and restart the connection with AT+CIPSTART. If the error still occurs,
it is recommended to shut off the PDP with AT+CIPSHUT and restart the connection,
if the error still persists, it is recommended to restart the module.


ChangeLog
2018-09-25	Evangelos Rokas <evrokas@gmail.com>
* Makefile: added git-upl
* error.[ch]: some progress on error logging and handling
* gsm.cpp: gsm_sendcmd*(): added parameter 'debug' default false
* main.cpp: transmitEEPROMstr(): added parameter 'debug' default false,
					setEEPROMstr(): in case E_SIMICCID it was setting addrSIMPIN, changed
					to addrSIMICCID instead,
					use __save_tail_pointer to save __tail_db pointer during POST method,
					if POST fails, then restore __tail_db in order to preserve records,
* posthttp.cpp: use _total_chunk_size to store total number of payload send to
					the server,
					http_send_post(): moved gsm_getBattery and gsm_getSignalQuality to
					the beginning of the function, so not to delay while sending data
					added \r\n in all data for the header,

2018-09-21	Evangelos Rokas <evrokas@gmail.com>
* posthttp.cpp: changed CRLF to LF, changed POST request to post.php
* 

2018-04-15	Evangelos Rokas <evrokas@gmail.com>
* added maintenance mode, documentation in CONFIG file
* Makefile: added -V -v -v in avrdude command line (more output, no verification)
					added NEOSW_RX_BUFFER to 32 instead of 64 which is the default (source changed)
* main.cpp: added EEPROM variables for various info that must be stored during restart,
						these are saved in the internal EEPROM of the chip,
						added moduleAflags which is a 16 bit word which holds, module enabled,
						DNS lookup enabled, etc
					(loadVariabledFromEEPROM): load some variables from EEPROM, maxMinLogCycle,
														maxMinNetCycle, Vcc factorial
* gsm.cpp: changed many calls to their respective *p to make use of the F() directive
* mem.cpp: mem_storecounter(), mem_readcounter() to store/read db counters to/from the
					RTC NVRAM (in the case of DS1307) or in the alarm0 register (in the case of DS3231)

2018-04-06	Evangelos Rokas <evrokas@gmail.com>
* added conf.h
* main.cpp: doMaintanance(): fix command line loop
* main(): add cnt32 variable to wait for a certain number of
			of Serial.available() calls, in order to sync with incoming
			'+' message
			add of APN, USERNAME, PASSWORD and DBSERVER defines (ad-hoc, later
			they will be changed with variables sitting in the EEPROM memory,
			to allow for parameter changing via the GSM/GPRS or the Serial
			interface
* ver.h.in: added $VERSION_SHORT

2017-10-26  Evangelos Rokas <evrokas@gmail.com>
* ver.h: added version header file,
* data.h: added dt field in datablock_t,
* gsm.cpp (http_send_datablock): fixed errors,
* main.cpp: added doNet, doLog flags and modified code accordingly, fix registration code,
					to sense correctly, resolve server name to get IP address using GSM module
					commands

2017-03-19	Evangelos Rokas <evrokas@gmail.com>
* Makefile.top: added flag AVRDUDE and AVRDUDE_CONF to point at system avrdude program
* gsm.cpp: new compile defines DEF_CLEAR_TEMPBUF and READGSM, use these to
		define, use and read to a temporary buffer with a specified length,
		added CF compile define
		added gsm_relayOutput(), fixed gsm_activateBearerProfile(),
		gsm_getBearerStatus(), gsm_deactivateBearerProfile()
		changed gsm_getRequest() to accept data length for received data
		gsm_getBattery() new implementation, added gsm_flushInput(), added
		gsm_sendPin(), gsm_moduleRead(), gsm_sendrecvcmdtimeout(),
		gsm_getTimeDateLonLat()
* gsm/main.c: added command for '*' and '^'
* utils.c: added powerRTC() function
		
2017-03-15	Evangelos Rokas <evrokas@gmail.com>
* data.h: mcuTemp is deprecated, commented out, changed designation of gpsLat,gpsLon
					to float and rtcDateTime to uint32_t (unix time)
* data.cpp: added gsm_moduleInfo(), gsm_getBattery() and gms_flushinput()
					functions
* gsm/main.cpp: some changes
* main.cpp: added code to perform network communications and handle basic
					module synchronization error

2017-03-14	Evangelos Rokas <evrokas@gmail.com>
* Makefile: added conditional define in CPPFLAGS, -DGSM_SERIAL_BAUDRATE
* gsm.h: make use of GSM_SERIAL_BAUDRATE and init it if not initialized on commandline
* gsm/main.cpp: some progress, added sapbr_init() and http_send() functions
  loop(): various cosmetic changes, changed to switch/case statements

2017-03-13	Evangelos Rokas <evrokas@gmail.com>
* Makefile.top: added AVRDUDE_ARG_PROGRAMMER option in order to select
                correct programmer. There are two boards defined in boards.txt
								bb8 and bb8isp. bb8 uses stk500v1 protocol and bb8isp uses
                arduinoisp protocol for avrdude upload

2017-02-14	Evangelos Rokas <evrokas@gmail.com>
* renamed main.ino to main.cpp

2017-02-12	Evangelos Rokas <evrokas@gmail.com>
* started adding F() function to place strings in FLASH memory
* added gsm.cpp and gsm.h, to start implementing GSM/GPRS functionality

2017-02-10	Evangelos Rokas <evrokas@gmail.com>
* README: added info for project reactivation

2017-02-09	Evangelos Rokas <evrokas@gmail.com>
* added ChangeLog in git repository
* added Makefile.top in git repository
* Makefile: added header preample, created Makefile.top and moved there some
  variables common to the project and the subprojects
* bms.h: added define SPEED_RATIO in order to simplify testing
* eeprom/Makefile: added include directive to Makefile.top
* main.ino: setup() moved initialization of maxSleepCycle 3 lines later
  added some debug ifdef/endifs in setup() and loop(),
  IMPORTANT: added Wire.begin() just before rtc_getTime()

2017-02-09	Evangelos Rokas <evrokas@gmail.com>
* Initial ChangeLog entry (9/2/2017)


pstr.h

/*
 * pstr.h - program memory strings
 *
 * Beehive Monitoring Project - BeeWatch
 *
 * (c) Evangelos Rokas, (c) 2018
 */

#ifndef __PSTR_H__
#define __PSTR_H__

#include <avr/pgmspace.h>
#include "bms.h"


#define RCF(str)	reinterpret_cast<const __FlashStringHelper *>(str)

const prog_char pATCIPSHUTrn [] PROGMEM	= "AT+CIPSHUT\r\n";
const prog_char pSHUTOK [] PROGMEM			= "SHUT OK";
const prog_char pCONNECT [] PROGMEM			= "CONNECT";
const prog_char pSEND [] PROGMEM				= "SEND";

const prog_char pATCIPMUX0rn [] PROGMEM	= "AT+CIPMUX=0\r\n";
const prog_char pOK [] PROGMEM					= "OK";
const prog_char pATSAPBR31CONTYPE [] PROGMEM		= "AT+SAPBR=3,1,\"CONTYPE\",\"GPRS\"\r\n";
const prog_char pATSAPBR31APN [] PROGMEM				= "AT+SAPBR=3,1,\"APN\",\"";
const prog_char pATSAPBR31USER [] PROGMEM				= "AT+SAPBR=3,1,\"USER\",\"";
const prog_char pATSAPBR31PWD [] PROGMEM				= "AT+SAPBR=3,1,\"PWD\",\"";
const prog_char pATSAPBR11rn [] PROGMEM					= "AT+SAPBR=1,1\r\n";
const prog_char pATSAPBR21rn [] PROGMEM					= "AT+SAPBR=2,1\r\n";
const prog_char pATSAPBR01rn [] PROGMEM					= "AT+SAPBR=0,1\r\n";

const prog_char pATCSTTQ [] PROGMEM							= "AT+CSTT=\"";
const prog_char pQcommaQ [] PROGMEM							= "\",\"";


const prog_char pATCIPSTATUSrn [] PROGMEM				= "AT+CIPSTATUS\r\n";
const prog_char pIPSTART [] PROGMEM							= "IP START";
const prog_char pATCIICRrn [] PROGMEM						= "AT+CIICR\r\n";
const prog_char pIPGPRSACT [] PROGMEM						= "IP GPRSACT";
const prog_char pATCIFSRrn [] PROGMEM						= "AT+CIFSR\r\n";
const prog_char pATCIPCLOSE [] PROGMEM					= "AT+CIPCLOSE\r\n";
const prog_char pCLOSEOK [] PROGMEM							= "CLOSE OK";

const prog_char pDOT [] PROGMEM									= ".";
const prog_char pCOMMA [] PROGMEM								= ",";
const prog_char pDDOT [] PROGMEM								= ":";
const prog_char pQ [] PROGMEM										= "\"";
const prog_char pAmp [] PROGMEM										= "&";
const prog_char pEq [] PROGMEM									= "=";
const prog_char pQddQ [] PROGMEM								= "\":\"";
const prog_char pCURLOPEN [] PROGMEM						= "{";
const prog_char pCURLCLOSE [] PROGMEM						= "}";
const prog_char pCRLF [] PROGMEM								= "\r\n";
const prog_char pCtrlZ [] PROGMEM								= "\032";

const prog_char pATCDNSGIPQ [] PROGMEM					= "AT+CDNSGIP=\"";
const prog_char pCDNSGIP [] PROGMEM							= "+CDNSGIP:";

const prog_char pATCIPSTART [] PROGMEM					= "AT+CIPSTART=\"TCP\",\"";
const prog_char pATCIPSEND [] PROGMEM						= "AT+CIPSEND\r\n";

const prog_char pATHTTPINITrn [] PROGMEM				= "AT+HTTPINIT\r\n";
const prog_char pATHTTPPARACIDrn [] PROGMEM			= "AT+HTTPPARA=\"CID\", 1\r\n";
const prog_char pATHTTPTERMrn [] PROGMEM				= "AT+HTTPTERM\r\n";
const prog_char pATHTTPPARAUArn [] PROGMEM			= "AT+HTTPPARA=\"UA\",\"BEEHIVE SIM MODULE\"\r\n";
const prog_char pATHTTPPARAURLQ [] PROGMEM			= "AT+HTTPPARA=\"URL\",\"http://";

const prog_char pGET [] PROGMEM				= "GET ";
const prog_char pPOST [] PROGMEM			= "POST ";
const prog_char pHTTP11 [] PROGMEM		= " HTTP/1.1\n";
const prog_char pHOST [] PROGMEM			= "Host: ";
const prog_char pUserAgent [] PROGMEM	= "User-Agent: beewatch/0.1 arduino/181\n";

const prog_char pACTION [] PROGMEM		= "action";
const prog_char pGETCONF [] PROGMEM		= "getconf";

const prog_char pDAT [] PROGMEM						= "dat";
const prog_char pGSM [] PROGMEM						= "gsm";
const prog_char PGPS [] PROGMEM						= "gps";


#if defined( API_STRING_KEYS )

//const prog_char pAPIKEY [] PROGMEM				= "apikey";
const prog_char pnodeId [] PROGMEM				= "nodeId";
const prog_char pentryType [] PROGMEM			= "entryType";
const prog_char pbatVolt [] PROGMEM				= "batVolt";
const prog_char pbhvTemp [] PROGMEM				= "bhvTemp";
const prog_char pbhvHumid [] PROGMEM			= "bhvHumid";
const prog_char pbhvWeight [] PROGMEM			= "bhvWeight";
const prog_char pgsmSig [] PROGMEM				= "gsmSig";
const prog_char pgsmVolt [] PROGMEM				= "gsmVolt";
const prog_char pgsmPDur [] PROGMEM				= "gsmPDur";
const prog_char pgpsLon [] PROGMEM				= "gpsLon";
const prog_char pgpsLat [] PROGMEM				= "gpsLat";
const prog_char prtcDateTime [] PROGMEM		= "rtcDateTime";

#elif defined( API_NUMERIC_KEYS )

//const prog_char pAPIKEY [] PROGMEM				= "0";			//"apikey";
const prog_char pnodeId [] PROGMEM				= "1";			//"nodeId";
const prog_char pentryType [] PROGMEM			= "2";			//"entryType";
const prog_char pbatVolt [] PROGMEM				= "3";			//"batVolt";
const prog_char pbhvTemp [] PROGMEM				= "4";			//"bhvTemp";
const prog_char pbhvHumid [] PROGMEM			= "5";			//"bhvHumid";
const prog_char pbhvWeight [] PROGMEM			= "6";			//"bhvWeight";
const prog_char pgsmSig [] PROGMEM				= "7";			//"gsmSig";
const prog_char pgsmVolt [] PROGMEM				= "8";			//"gsmVolt";
const prog_char pgsmPDur [] PROGMEM				= "9";			//"gsmPDur";
const prog_char pgpsLon [] PROGMEM				= "10";			//"gpsLon";
const prog_char pgpsLat [] PROGMEM				= "11";			//"gpsLat";
const prog_char prtcDateTime [] PROGMEM		= "12";			//"rtcDateTime";

#else

#error	Please define API_STRING_KEYS or API_NUMERIC_KEYS in bms.h

#endif



const prog_char pATHTTPACTION0rn [] PROGMEM	= "AT+HTTPACTION=0\r\n";
const prog_char pHTTPACTION [] PROGMEM			= "+HTTPACTION:";
const prog_char pATHTTPREADrn [] PROGMEM 		= "AT+HTTPREAD\r\n";
const prog_char pHTTPREAD [] PROGMEM				= "+HTTPREAD: ";

const prog_char pATIrn [] PROGMEM				= "ATI\r\n";
const prog_char pATCPINeq [] PROGMEM		= "AT+CPIN=";
const prog_char pATrn [] PROGMEM				= "AT\r\n";
const prog_char pRN [] PROGMEM					= "\r\n";
const prog_char pLF [] PROGMEM						= "\n";

const prog_char pATCCIDrn [] PROGMEM		= "AT+CCID\r\n";
const prog_char pATCREGqrn [] PROGMEM		= "AT+CREG?\r\n";
//const prog_char pCREG [] PROGMEM				= "+CREG: ";
const prog_char pATCFUNC0rn [] PROGMEM	= "AT+CFUNC=0\r\n";
const prog_char pATCFUNC1rn [] PROGMEM	= "AT+CFUNC=1\r\n";
const prog_char pATCBCrn [] PROGMEM			= "AT+CBC\r\n";
const prog_char pCBC [] PROGMEM					= "+CBC: ";
const prog_char pATCSQrn [] PROGMEM			= "AT+CSQ\r\n";
const prog_char pCSQ [] PROGMEM					= "+CSQ: ";
const prog_char pATCIPGSMLOC11rn [] PROGMEM	= "AT+CIPGSMLOC=1,1\r\n";
const prog_char pCIPGSMLOC [] PROGMEM					= "+CIPGSMLOC:";		/* take care not to include the last space */
const prog_char pATCIPGSMLOC21rn [] PROGMEM	= "AT+CIPGSMLOC=2,1\r\n";

const prog_char pNoReceivedLength [] PROGMEM	= "no received bytes length";
const prog_char pErrorCPIN [] PROGMEM					= "error setting SIM pin";

const prog_char pQrn [] PROGMEM			= "\"\r\n";

const prog_char pNodeId [] PROGMEM	= "Node ID: ";
const prog_char pWrongParameter [] PROGMEM = "wrong parameter";
const prog_char pErrorCouldNot [] PROGMEM = "error: could not ";
const prog_char pError [] PROGMEM = "error: ";
const prog_char pErrorSendDATblock [] PROGMEM = "send data block";
const prog_char pErrorSendGSMblock [] PROGMEM = "send gsm block";
const prog_char pErrorInitiateGet [] PROGMEM = "initiate get request";
const prog_char pErrorActivateBearer [] PROGMEM ="activate bearer profile";
const prog_char pErrorInitiatePost [] PROGMEM = "initiate post request";
const prog_char pFailedPostSend [] PROGMEM = "failed POST send";

#endif	/* __PSTR_H__ */
